### PR TITLE
ISSUE-1045 Implement JMAP route to get content of the public asset

### LINKED ISSUE #1045: Implement JMAP route to get content of the public asset
## Why 
Epic: https://github.com/linagora/tmail-backend/issues/1027
When a client (e.g., web browser, mobile app) requests the content of a public asset, the Tmail JMAP server needs to serve that content.

## How

- Create a new `PublicAssetRoutes` class with the following implementation:

```scala
class PublicAssetRoutes(publicAssetRepository: PublicAssetRepository,
                        blobResolvers: BlobResolvers) extends JMAPRoutes {

  override def routes(): Stream[JMAPRoute] = Stream.of(
    JMAPRoute.builder()
      .endpoint(new Endpoint(HttpMethod.GET, s"/publicAsset/{$accountId}/{$assetId}"))
      .action((request, response) => getAsset(request, response))
      .corsHeaders())

  private def getAsset(request: HttpServerRequest, response: HttpServerResponse): SMono[Unit] = {
    // checking if accountId and assetId are existing
    // if not, return 404
    // else, using blobResolvers to get the blob of the asset
    // then apply the blob to the response
  }
}
```

Note that this is a public endpoint, so authentication is not required.

- Bind `PublicAssetRoutes` to `JMAPRoutes` using `Guice`:
```
Multibinder.newSetBinder(binder, classOf[JMAPRoutes])
    routes.addBinding().to(classOf[PublicAssetRoutes])
```

- Write integration tests.

## Dod 
- Tests pass.
- Documentation.

Ref: 
[LinagoraServicesDiscoveryRoutes](https://github.com/linagora/tmail-backend/blob/78af34f4700201b965e78b4651ae1993c69d22fb/tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/service/discovery/LinagoraServicesDiscoveryRoutes.scala#L75)

[DownloadRoutes](https://github.com/apache/james-project/blob/fd83dae0bf5b626c6d0276523307ec1816f48b3e/server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/DownloadRoutes.scala#L246)

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve #1045

### CODE REVIEW

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+      case _ => returnNotFound(response)
+          .asJava()
+          .`then`
+    }
+  }
+
+  private def getAsset(request: HttpServerRequest, response: HttpServerResponse, username: Username, assetId: PublicAssetId): Mono[Void] =
+    SMono(publicAssetRepository.get(username, assetId))
+      .switchIfEmpty(SMono.error(PublicAssetNotFoundException(assetId)))
+      .flatMap(asset => SMono(blobStore.readReactive(blobStore.getDefaultBucketName, asset.blobId, StoragePolicy.LOW_COST))
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala:72]: Why does the blobstore, a storage detail, leak into the JMAP presentation layer?

I would expect the repository to encapsulate such details...

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        case e =>
+          LOGGER.error("Unexpected error upon downloading public asset {}", request.uri(), e)
+          respondDetails(response, ProblemDetails(status = INTERNAL_SERVER_ERROR, detail = e.getMessage), INTERNAL_SERVER_ERROR)
+      }
+      .subscribeOn(ReactorUtils.BLOCKING_CALL_WRAPPER)
+      .asJava()
+      .`then`
+
+  private def returnNotFound(response: HttpServerResponse): SMono[Unit] =
+    respondDetails(response, ProblemDetails(status = NOT_FOUND, detail = "The public asset could not be found"), NOT_FOUND)
```
Comment (chibenwa) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala:88]: Log info?

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala (hunk 1 above)
Comment (quantranhong1999) [reply tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala]: Should not actually (I relied on the old repository API).

**Review by Arsnael — APPROVED** [review]

**Review by hungphan227 — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      .asJava()
+      .`then`
+
+  private def returnNotFound(request: HttpServerRequest, response: HttpServerResponse): SMono[Unit] = {
+    LOGGER.info("Can not serve not found public asset {}", request.uri())
+    respondDetails(response, ProblemDetails(status = NOT_FOUND, detail = "The public asset could not be found"), NOT_FOUND)
+  }
+
+  private def downloadAsset(response: HttpServerResponse,
+                           contentType: ContentType,
```
Comment (vttranlina) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala:89]: space
=> Reformat will be pretty

**Review by vttranlina — APPROVED** [review]

File: tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                           content: InputStream,
+                           size: Size): SMono[Unit] =
```
Comment (quantranhong1999) [inline tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala:91]:
```suggestion
                            contentType: ContentType,
                            content: InputStream,
                            size: Size): SMono[Unit] =
```

**Comment by Arsnael** [comment]
Needs rebase

**Pull request merged**